runtime/queries/metricsview.go

```diff
@@ -0,0 +1,312 @@
+package queries
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"math/big"
+	"strings"
+	"time"
+	"unicode/utf8"
+
+	"github.com/marcboeker/go-duckdb"
+	runtimev1 "github.com/rilldata/rill/proto/gen/rill/runtime/v1"
+	"github.com/rilldata/rill/runtime"
+	"github.com/rilldata/rill/runtime/drivers"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+func lookupMetricsView(ctx context.Context, rt *runtime.Runtime, instanceID string, name string) (*runtimev1.MetricsView, error) {
+	obj, err := rt.GetCatalogEntry(ctx, instanceID, name)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	if obj.GetMetricsView() == nil {
+		return nil, status.Errorf(codes.NotFound, "object named '%s' is not a metrics view", name)
+	}
+
+	return obj.GetMetricsView(), nil
+}
+
+func metricsQuery(ctx context.Context, olap drivers.OLAPStore, priority int, sql string, args []any) ([]*runtimev1.MetricsViewColumn, []*structpb.Struct, error) {
+	rows, err := olap.Execute(ctx, &drivers.Statement{
+		Query:    sql,
+		Args:     args,
+		Priority: priority,
+	})
+	if err != nil {
+		return nil, nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+	defer rows.Close()
+
+	data, err := rowsToData(rows)
+	if err != nil {
+		return nil, nil, status.Error(codes.Internal, err.Error())
+	}
+
+	return structTypeToMetricsViewColumn(rows.Schema), data, nil
+}
+
+func rowsToData(rows *drivers.Result) ([]*structpb.Struct, error) {
+	var data []*structpb.Struct
+	for rows.Next() {
+		rowMap := make(map[string]any)
+		err := rows.MapScan(rowMap)
+		if err != nil {
+			return nil, err
+		}
+
+		rowStruct, err := mapToPB(rowMap)
+		if err != nil {
+			return nil, err
+		}
+
+		data = append(data, rowStruct)
+	}
+
+	err := rows.Err()
+	if err != nil {
+		return nil, err
+	}
+
+	return data, nil
+}
+
+func structTypeToMetricsViewColumn(v *runtimev1.StructType) []*runtimev1.MetricsViewColumn {
+	res := make([]*runtimev1.MetricsViewColumn, len(v.Fields))
+	for i, f := range v.Fields {
+		res[i] = &runtimev1.MetricsViewColumn{
+			Name:     f.Name,
+			Type:     f.Type.Code.String(),
+			Nullable: f.Type.Nullable,
+		}
+	}
+	return res
+}
+
+// valToPB converts any value to a google.protobuf.Value. It's similar to
+// structpb.NewValue, but adds support for a few extra primitive types.
+func valToPB(v any) (*structpb.Value, error) {
```

Comment:
looks like it belongs to protobuf util ? and might be reused?

Reply:
Yes, this is now replicated here and in `runtime/server/query.go`. Maybe put it (and related functions) in `runtime/server/pbutil` as `ToValue(v any) (*structpb.Value, error)`?